File: maheshwari_motors_app/lib/app/data/models/challan_model.dart
import 'challan_item_model.dart';

class ChallanModel {
  final String id;
  final int? numericId;
  final String challanNo;
  final DateTime date;
  final String? partyId;
  final String? partyName;
  final List<ChallanItemModel> items;
  final double grossTotal;
  final double subTotal;
  final double amount;
  final double discount;
  final bool convertedToBill;
  final String? billId;
  final int isGst;
  final String? linkedChallanId;

  ChallanModel({
    required this.id,
    this.numericId,
    required this.challanNo,
    required this.date,
    this.partyId,
    this.partyName,
    this.items = const [],
    this.grossTotal = 0,
    this.subTotal = 0,
    required this.amount,
    this.discount = 0,
    this.convertedToBill = false,
    this.billId,
    this.isGst = 1,
    this.linkedChallanId,
  });

  factory ChallanModel.fromJson(Map<String, dynamic> json) {
    final partyData = json['party_id'];
    return ChallanModel(
      id: json['_id'] ?? '',
      numericId: json['id'],
      challanNo: json['challan_no'] ?? '',
      date: DateTime.tryParse(json['date'] ?? '') ?? DateTime.now(),
      partyId: partyData is Map ? partyData['_id'] : partyData?.toString(),
      partyName: partyData is Map ? partyData['name'] : null,
      items: json['items'] != null
          ? (json['items'] as List)
                .map((e) => ChallanItemModel.fromJson(e))
                .toList()
          : [],
      grossTotal: (json['gross_total'] ?? 0).toDouble(),
      subTotal: (json['sub_total'] ?? 0).toDouble(),
      amount: (json['amount'] ?? 0).toDouble(),
      discount: (json['discount'] ?? 0).toDouble(),
      convertedToBill: json['converted_to_bill'] ?? false,
      billId: json['bill_id']?.toString(),
      isGst: json['is_gst'] ?? 1,
      linkedChallanId: json['linked_challan_id']?.toString(),
    );
  }

  Map<String, dynamic> toJson() {
    final map = <String, dynamic>{
      '_id': id,
      'challan_no': challanNo,
      'date': date.toIso8601String(),
      'items': items.map((e) => e.toJson()).toList(),
      'gross_total': grossTotal,
      'sub_total': subTotal,
      'amount': amount,
      'discount': discount,
      'converted_to_bill': convertedToBill,
      'is_gst': isGst,
    };
    if (partyId != null) map['party_id'] = partyId;
    if (billId != null) map['bill_id'] = billId;
    if (linkedChallanId != null) map['linked_challan_id'] = linkedChallanId;
    return map;
  }
}

File: backend/src/models/challan.model.js
import mongoose from "mongoose";

const challanSchema = new mongoose.Schema(
  {
    challan_no: { type: String, required: true },
    date: { type: Date, required: true, default: Date.now },
    party_id: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "Party",
      required: true,
    },
    items: [
      {
        item_id: {
          type: mongoose.Schema.Types.ObjectId,
          ref: "Item",
          required: true,
        },
        quantity: { type: Number, default: 1 },
        rate: { type: Number, required: true },
        discount: { type: Number, default: 0 },
        gross_amount: { type: Number, required: true },
        amount: { type: Number, required: true },
        is_gst: { type: Number, enum: [0, 1], default: 1 },
      },
    ],
    gross_total: { type: Number, required: true },
    sub_total: { type: Number, required: true },
    discount: { type: Number, default: 0 },
    amount: { type: Number, required: true },
    converted_to_bill: { type: Boolean, default: false },
    bill_id: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "Bill",
      default: null,
    },

    // Challan-level GST flag: 1 = GST challan (under GST firm), 0 = NON_GST challan
    is_gst: { type: Number, enum: [0, 1], required: true },

    // Links the twin challan created from the same submission (auto-split)
    linked_challan_id: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "Challan",
      default: null,
    },

    user_id: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "User",
    },
    firm_id: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "Firm",
      required: true,
    },
  },
  { timestamps: true },
);

challanSchema.index({ challan_no: 1, firm_id: 1 }, { unique: true });

export default mongoose.model("Challan", challanSchema);
